### PR TITLE
Activate filesystem features only in syncing context

### LINKED ISSUE #14252: zfs-2.1.7 zfs recv panic in  dsl_dataset_deactivate_feature_impl
### System information

Type | Version/Name
 --- | ---
Distribution Name	| Fedora
Distribution Version	| 36
Kernel Version	| 6.0.10-200.fc36.x86_64
Architecture	| Intel x86_64
OpenZFS Version	| zfs-2.1.7-1

### Describe the problem you're observing
When using zfs send to make a backup on a remote machine, the receiver throws a PANIC in one of the zfs functions and the file system deadlocks.

### Describe how to reproduce the problem
```
zfs send -RLec -I snapshot_name | ssh remote_host "zfs receive -duF remote_zvol"
```

### Include any warning/errors/backtraces from the system logs
```
System Events
=-=-=-=-=-=-=
Dec  2 15:32:14 raid3 kernel: VERIFY3(0 == zap_remove(mos, dsobj, spa_feature_table[f].fi_guid, tx)) failed (0 == 2)
Dec  2 15:32:14 raid3 kernel: PANIC at dsl_dataset.c:1116:dsl_dataset_deactivate_feature_impl()
Dec  2 15:32:14 raid3 kernel: Showing stack for process 1604
Dec  2 15:32:14 raid3 kernel: CPU: 3 PID: 1604 Comm: txg_sync Tainted: P           OE      6.0.10-200.fc36.x86_64 #1
Dec  2 15:32:14 raid3 kernel: Hardware name: Gigabyte Technology Co., Ltd. Z87X-UD3H/Z87X-UD3H-CF, BIOS 10b 06/12/2014
Dec  2 15:32:14 raid3 kernel: Call Trace:
Dec  2 15:32:14 raid3 kernel: <TASK>
Dec  2 15:32:14 raid3 kernel: dump_stack_lvl+0x44/0x5c
Dec  2 15:32:14 raid3 kernel: spl_panic+0xf0/0x108 [spl]
Dec  2 15:32:14 raid3 kernel: dsl_dataset_deactivate_feature_impl+0xfb/0x100 [zfs]
Dec  2 15:32:14 raid3 kernel: dsl_dataset_clone_swap_sync_impl+0x88e/0xa70 [zfs]
Dec  2 15:32:14 raid3 kernel: ? _raw_spin_unlock+0x15/0x30
Dec  2 15:32:14 raid3 kernel: ? dsl_dataset_hold_flags+0xb0/0x230 [zfs]
Dec  2 15:32:14 raid3 kernel: dmu_recv_end_sync+0x180/0x560 [zfs]
Dec  2 15:32:14 raid3 kernel: ? preempt_count_add+0x6a/0xa0
Dec  2 15:32:14 raid3 kernel: dsl_sync_task_sync+0xa5/0xf0 [zfs]
Dec  2 15:32:14 raid3 kernel: dsl_pool_sync+0x3d3/0x4e0 [zfs]
Dec  2 15:32:14 raid3 kernel: spa_sync+0x555/0xf60 [zfs]
Dec  2 15:32:14 raid3 kernel: ? _raw_spin_unlock+0x15/0x30
Dec  2 15:32:14 raid3 kernel: ? spa_txg_history_init_io+0xf3/0x110 [zfs]
Dec  2 15:32:14 raid3 kernel: txg_sync_thread+0x227/0x3e0 [zfs]
Dec  2 15:32:14 raid3 kernel: ? txg_fini+0x260/0x260 [zfs]
Dec  2 15:32:14 raid3 kernel: ? __thread_exit+0x20/0x20 [spl]
Dec  2 15:32:14 raid3 kernel: thread_generic_wrapper+0x57/0x70 [spl]
Dec  2 15:32:14 raid3 kernel: kthread+0xe6/0x110
Dec  2 15:32:14 raid3 kernel: ? kthread_complete_and_exit+0x20/0x20
Dec  2 15:32:14 raid3 kernel: ret_from_fork+0x1f/0x30
Dec  2 15:32:14 raid3 kernel: </TASK>
```

### PR DESCRIPTION
**Pull request by gamanakis**
### Motivation and Context

Closes #14252 

### Description

When activating filesystem features after receiving a snapshot, do so only in syncing context.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: module/zfs/dsl_dataset.c [hunk 1]
```diff
@@ -1761,16 +1761,19 @@ dsl_dataset_snapshot_sync_impl(dsl_dataset_t *ds, const char *snapname,
 
 	/*
 	 * We are not allowed to dirty a filesystem when done receiving
-	 * a snapshot. In this case the flag SPA_FEATURE_LARGE_BLOCKS will
-	 * not be set and a subsequent encrypted raw send will fail. Hence
-	 * activate this feature if needed here.
+	 * a snapshot. In this case some flags such as SPA_FEATURE_LARGE_BLOCKS
+	 * will not be set and a subsequent encrypted raw send will fail. Hence
+	 * activate this feature if needed here. This needs to happen only in
+	 * syncing contect.
```
Comment (pharshalp) [inline module/zfs/dsl_dataset.c:1767]:
```suggestion
	 * syncing context.
```

**Review by behlendorf — APPROVED** [review]

**Review by freqlabs — APPROVED** [review]

**Comment by gamanakis** [comment]
We probably should restrict this piece of code to be executed only in case of raw receiving. Looking into it.

**Comment by gamanakis** [comment]
Restricting to raw receiving doesn't help the original case: The problem arises already when the receive is done and the feature flags are not set prior to raw sending.

**Comment by gamanakis** [comment]
@behlendorf this patch does not address all the issues with #14252. I am looking into it.

**Comment by gamanakis** [comment]
I updated the PR to address all the issues with #14252. I do not think I introduced any new regressions.

**Comment by gamanakis** [comment]
@rincebrain would you mind me adding the test you provided in #14252 here?

**Comment by rincebrain** [comment]
By all means, go for it. I'm sure there's a simpler test case that works, but I haven't spent any time trying to minimize it, since I wasn't even...trying to make it when I did. Heh.

**Review by behlendorf — APPROVED** [review]

**Review by ryao — APPROVED** [review]

File: tests/zfs-tests/tests/functional/rsend/send_encrypted_incremental.ksh [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+typeset file2="/pool_lb/encryptme/$TESTFILE2"
+
+# Create pool
+truncate -s $MINVDEVSIZE $TESTDIR/vdev_a
+log_must zpool create -f pool_lb $TESTDIR/vdev_a
+log_must eval "echo $passphrase > /pool_lb/pwd"
+log_must eval "echo $passphrase2 > /pool_lb/pwd2"
+
+log_must eval "zfs create pool_lb/encryptme -o encryption=on \
+	-o keyformat=passphrase	-o keylocation=file:///pool_lb/pwd"
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/rsend/send_encrypted_incremental.ksh:71]: There's no need to use `eval` here, this is probably responsible for this test failing in the CI.  

http://build.zfsonlinux.org/builders/FreeBSD%20stable%2F13%20amd64%20%28TEST%29/builds/7583/steps/shell_4/logs/summary

File: tests/zfs-tests/tests/functional/rsend/send_encrypted_incremental.ksh [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+log_must dd if=/dev/urandom of=$file bs=1024 count=1024
+log_must zfs snapshot pool_lb/encryptme@snap1
+log_must dd if=/dev/urandom of=$file1 bs=1024 count=1024
+log_must zfs snapshot pool_lb/encryptme@snap2
+log_must dd if=/dev/urandom of=$file2 bs=1024 count=1024
+log_must zfs snapshot pool_lb/encryptme@snap3
+log_must eval "zfs send -Lc pool_lb/encryptme@snap1 | sudo zfs recv \
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/rsend/send_encrypted_incremental.ksh:78]:
```suggestion
log_must eval "zfs send -Lc pool_lb/encryptme@snap1 | zfs recv \
```

Not need for `sudo` throughout.  These test are run with escalated privileges.

**Comment by gamanakis** [comment]
@behlendorf should be an easy fix, I think it doesn't like the order of the arguments. I will get to it later today.

**Comment by gamanakis** [comment]
done!

**Comment by behlendorf** [comment]
Much better, thanks!

**Comment by peterska** [comment]
Will this be backported to 2.1.8 ?

**Comment by ryao** [comment]
> Will this be backported to 2.1.8 ?

As far as I know, yes.

**Pull request merged**